Sources/Crust/Parser.swift

```diff
@@ -42,4 +50,543 @@ public class Parser {
   func advance(_ n: Int = 1) {
     index += n
   }
+
+  func split<T>(_ fs : (() throws -> T)...) throws -> T {
+    for f in fs {
+      let previousPosition = self.index
+      guard let syntax = try? f() else {
+        self.index = previousPosition
+        continue
+      }
+      return syntax
+    }
+    throw ParseError.allDisjunctsFailed
+  }
+}
+
+extension Parser {
+  public func parseTopLevelModule() -> ModuleDeclSyntax? {
+    let module = try? parseModule()
+    guard let _ = try? consume(.eof) else {
+      return nil
+    }
+    return module
+  }
+}
+
+extension Parser {
+  func parseQualifiedName() throws -> QualifiedNameSyntax {
+    var pieces = [QualifiedNamePieceSyntax]()
+    while let piece = try? parseQualifiedNamePiece() {
+      pieces.append(piece)
+    }
+
+    guard !pieces.isEmpty else {
+      throw ParseError.failedParsingName
+    }
+
+    return QualifiedNameSyntax(elements: pieces)
+  }
+
+  func parseQualifiedNamePiece() throws -> QualifiedNamePieceSyntax {
+    guard case .identifier(_) = self.peek() else {
+      throw ParseError.unexpectedToken(self.currentToken!)
+    }
+
+    let name = self.currentToken!
+    self.advance()
+    let period = (peek() == .period) ? try? consume(.period) : nil
+    return QualifiedNamePieceSyntax(name: name, trailingPeriod: period)
+  }
+
+  func parseIdentifierList() throws -> IdentifierListSyntax {
+    var pieces = [TokenSyntax]()
+    while let piece = try? parseQualifiedNamePiece() {
+      pieces.append(piece.name)
+    }
+    return IdentifierListSyntax(elements: pieces)
+  }
+}
+
+extension Parser {
+  func parseDeclList() -> DeclListSyntax {
+    var pieces = [DeclSyntax]()
+    while let piece = try? parseDecl() {
+      pieces.append(piece)
+    }
+    return DeclListSyntax(elements: pieces)
+  }
+
+  func parseDecl() throws -> DeclSyntax {
+    return try split({
```

Comment:
Why not switch on the token kind here? It’d be much easier to understand and way faster...